src/Chapter4.hs

```diff
@@ -472,10 +499,12 @@ Implement the Applicative instance for our 'Secret' data type from before.
 -}
 instance Applicative (Secret e) where
     pure :: a -> Secret e a
-    pure = error "pure Secret: Not implemented!"
+    pure = Reward
 
     (<*>) :: Secret e (a -> b) -> Secret e a -> Secret e b
-    (<*>) = error "(<*>) Secret: Not implemented!"
+    Trap e <*> _ = Trap e
+    Reward _ <*> Trap b = Trap b
+    Reward a <*> Reward b = fmap a (Reward b)
```

Comment:
You can actually reduce these two cases to a single one. I see that you use `fmap`, and `fmap` already patterns matches on `Secret` and applies a function to it. So you can reuse `Functor` instance for `Secret` here